databricks_cli/tunnel/server.py

```diff
@@ -0,0 +1,142 @@
+
+
+class TunnelConfig:
+    def __init__(self, databricks_config, org_id, cluster_id):
+        self.host = databricks_config.host
+        self.token = databricks_config.token
+        self.org_id = org_id
+        self.cluster_id = cluster_id
+        self.remote_port = DBX_TUNNEL_SERVER_PORT
+        # TODO(tunneling-cli): this assumes a multitenant workspace, we will need to confirm if
+        #  this is still the case for st workspace
+        self.endpoint_path = \
+            f"driver-proxy-api/o/{self.org_id}/{self.cluster_id}/{self.remote_port}/"
+
+    @property
+    def endpoint_url(self):
+        return urljoin(self.host, self.endpoint_path)
+
+    @property
+    def healthcheck_url(self):
+        return f"{self.endpoint_url}health"
+
+    @property
+    def socket_path(self):
+        return f"{self.endpoint_path}socket.io"
+
+
+class TunnelClient:
+    def __init__(self, stream: IOStream, session: aiohttp.ClientSession,
+                 tunnel_config: TunnelConfig, debug=False):
+        self.stream = stream
+        self.tunnel_config = tunnel_config
+        self.io = socketio.AsyncClient(http_session=session, logger=debug, engineio_logger=debug)
+        self.connection_established = False
+
+        @self.io.event
+        def connect():  # pylint: disable=unused-variable
+            self.connection_established = True
+            logger.info("Connection to socket.io server established with id %s", self.io.sid)
+
+        @self.io.on("response")
+        def response(msg):  # pylint: disable=unused-variable
+            data = msg.get("data")
+            # receive responses from tunnel server then write to tcp's stream
+            if isinstance(data, bytes):
+                self.stream.write(data)
+            else:
+                logger.warning("Received string message from byte-based stream: %s", data)
+
+    async def connect(self):
+        while not self.io.connected:
+            logger.info("initializing socket.io connection")
+            try:
+                headers = {
+                    'Content-Type': 'application/json'
+                }
+                await self.io.connect(self.tunnel_config.endpoint_url,
+                                      auth=None,
+                                      transports=["websocket"],
+                                      socketio_path=self.tunnel_config.socket_path,
+                                      headers=headers)
+            except socketio.exceptions.ConnectionError as e:
+                logger.error("Connection exception:")
+                logger.error(e)
+                raise e
+
+        while not self.connection_established:
+            logger.info("Connection to the socket.io server is not yet established")
+            await asyncio.sleep(1)
+
+    def is_connected(self):
+        return self.io.connected
+
+    async def send_data(self, data):
+        await self.io.emit("data_received", {"data": data})
+
+    async def disconnect(self):
```

Comment:
@marcom-db, do we need to add an event handler for `disconnect` that sets `connection_established` to `False` as well?

Reply:
yeah, we can set it to false here if we disconnect